state_tracking.go

```diff
@@ -28,30 +30,53 @@ type StateTrackingNode struct {
 	as     string
 
 	newTracker func() stateTracker
-	groups     map[models.GroupID]*stateTrackingGroup
+
+	groupsMu sync.RWMutex
+	groups   map[models.GroupID]*stateTrackingGroup
 }
 
 func (stn *StateTrackingNode) group(g models.GroupID) (*stateTrackingGroup, error) {
+	stn.groupsMu.RLock()
 	stg := stn.groups[g]
-	if stg == nil {
-		stg = &stateTrackingGroup{}
+	stn.groupsMu.RUnlock()
 
-		var err error
-		stg.Expression, err = stateful.NewExpression(stn.lambda.Expression)
-		if err != nil {
-			return nil, fmt.Errorf("Failed to compile expression: %v", err)
-		}
+	if stg == nil {
+		// Grab the write lock
+		stn.groupsMu.Lock()
+		defer stn.groupsMu.Unlock()
+
+		// Check again now that we have the write lock
+		stg = stn.groups[g]
+		if stg == nil {
```

Comment:
Sorry, I know this is merged, but this is just bugging me.
Having this code implies that you can be processing 2 different points for the same node, and the same group at the same time. This isn't the case now, and I don't think it ever will be the case, because otherwise you'd have race conditions where the node might not emit points in the same order in which they came in.